### PR TITLE
18 gasometer

### PR DESCRIPTION
**Pull request by vakond**
Enabled gasometer. 

Note: return value in the proxy's function eth_estimateGas should be increased to pass tests, like so:
```Python
    def eth_estimateGas(self, param):
          if param.get('to'):
            call_emulated(param['to'], param['from'], param['data'])
        return 9999999
```

This change and other needed present in the branch 18-gasometer in the proxy-model.py repository.

### CODE REVIEW

File: evm_loader/program/src/entrypoint.rs [hunk 1]
```diff
@@ -484,22 +495,34 @@ fn do_partial_call<'a>(
...
+    let executor_state = ExecutorState::new(ExecutorSubstate::new(gas_limit, config), backend);
     let mut executor = Machine::new(executor_state);
 
     debug_print!("Executor initialized");
 
     debug_print!("   caller: {}", account_storage.origin());
     debug_print!(" contract: {}", account_storage.contract());
 
-    executor.call_begin(account_storage.origin(), account_storage.contract(), instruction_data, u64::max_value());
+    if executor.call_begin(
```
Comment (s-medvedev) [inline evm_loader/program/src/entrypoint.rs:515]: You can use simpler construction:
```
executor.call_begin(...).map_err(|_| ProgramError::InvalidInstructionData)?;

File: .gitmodules [hunk 2]
```diff
@@ -1,4 +1,4 @@
 [submodule "evm_loader/rust-evm"]
 	path = evm_loader/rust-evm
 	url = https://github.com/neonlabsorg/evm.git
-	branch = evm_integration
+	branch = 18-gasometer
```
Comment (s-medvedev) [inline .gitmodules:4]: Set `develop` branch here.

File: evm_loader/program/src/entrypoint.rs [hunk 3]
```diff
@@ -214,10 +212,10 @@ fn process_instruction<'a>(
             let mut storage = StorageAccount::new(storage_info, accounts, from_addr, trx.nonce)?;
 
             if trx.to.is_some() {
-                do_partial_call(&mut storage, step_count, &account_storage, accounts, trx.call_data)?;
+                do_partial_call(&mut storage, step_count, &account_storage, accounts, trx.call_data, trx.gas_limit.as_u64())?;
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs:215]: `as_u64()` can panic. Prefer `try_from/try_into` for conversions

File: evm_loader/program/src/entrypoint.rs [hunk 4]
```diff
@@ -395,11 +393,13 @@ fn do_finalize<'a>(program_id: &Pubkey, accounts: &'a [AccountInfo<'a>]) -> Prog
             code.to_vec()
         };
 
-        let executor_state = ExecutorState::new(ExecutorSubstate::new(), backend);
+        let gas_limit = u64::MAX;
+	    let config = evm::Config::default();
+        let executor_state = ExecutorState::new(ExecutorSubstate::new(gas_limit, config), backend);
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs:398]: Do not create config there. Use `Config::default()` inside substate constructor

File: evm_loader/program/src/entrypoint.rs (hunk 1 above)
Comment (vakond) [reply evm_loader/program/src/entrypoint.rs]: Good point, thanks.

File: evm_loader/program/src/executor.rs [hunk 5]
```diff
@@ -309,28 +318,39 @@ impl<'config, B: Backend> Machine<'config, B> {
         Self{ executor, runtime }
     }
 
-    pub fn call_begin(&mut self, caller: H160, code_address: H160, input: Vec<u8>, gas_limit: u64) {
+    pub fn call_begin(&mut self,
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor.rs:321]: You should record transaction cost as well.
See how it is done in the original rust-evm - https://github.com/rust-blockchain/evm/blob/7a225c38dd5e308fe634e6606cd2a330dc1e24cc/src/executor/stack/mod.rs#L257

File: evm_loader/program/src/executor.rs [hunk 6]
```diff
@@ -341,10 +361,10 @@ impl<'config, B: Backend> Machine<'config, B> {
         let runtime = evm::Runtime::new(code, input, context, self.executor.config);
 
         self.runtime.push((runtime, CreateReason::Call));
+        Ok(())
     }
 
     pub fn create_begin(&mut self, caller: H160, code: Vec<u8>, gas_limit: u64) -> ProgramResult {
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor.rs:367]: You must record gas for create transaction.
Also check how gas recorded for internal call and internal create.

File: evm_loader/program/src/executor.rs [hunk 7]
```diff
@@ -402,12 +421,12 @@ impl<'config, B: Backend> Machine<'config, B> {
...
     #[allow(clippy::too_many_lines)]
     pub fn step(&mut self) -> Result<(), ExitReason> {
-
+        let gas_limit = self.executor.state.block_gas_limit().as_u64();
         match self.step_opcode(){
             RuntimeApply::Continue => { Ok(()) },
             RuntimeApply::Call(info) => {
                 let code = self.executor.code(info.code_address);
-                self.executor.state.enter(u64::max_value(), false);
+                self.executor.state.enter(gas_limit, false);
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor.rs:429]: Why you are using `block_gas_limit()` there?

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

File: evm_loader/program/src/entrypoint.rs (hunk 3 above)
Comment (vakond) [reply evm_loader/program/src/entrypoint.rs]: Excellent point. Reworked to u64::try_from here and everywhere as_u64 had been used.

File: evm_loader/program/src/executor.rs (hunk 7 above)
Comment (vakond) [reply evm_loader/program/src/executor.rs]: Wasn't sure if the block gas limit makes sense here. Reverted to u64::MAX.

File: evm_loader/program/src/entrypoint.rs (hunk 4 above)
Comment (vakond) [reply evm_loader/program/src/entrypoint.rs]: Good point. Moved Config creation there, and the code became more concise and readable.

File: evm_loader/program/src/executor.rs (hunk 5 above)
Comment (vakond) [reply evm_loader/program/src/executor.rs:321]: A clarification is needed. We already have such code in our evm (branch 18-gasometer, of course): https://github.com/neonlabsorg/evm/blob/5b72bc771ef85750ad7fff056922a137592b38ea/src/executor/stack.rs#L212

Comment (vakond) [reply evm_loader/program/src/executor.rs:321]: > You should record transaction cost as well.
> See how it is done in the original rust-evm - https://github.com/rust-blockchain/evm/blob/7a225c38dd5e308fe634e6606cd2a330dc1e24cc/src/executor/stack/mod.rs#L257

You should record transaction cost as well.
See how it is done in the original rust-evm - https://github.com/rust-blockchain/evm/blob/7a225c38dd5e308fe634e6606cd2a330dc1e24cc/src/executor/stack/mod.rs#L257

File: evm_loader/program/src/executor.rs (hunk 6 above)
Comment (vakond) [reply evm_loader/program/src/executor.rs:367]: Well, it's done: https://github.com/neonlabsorg/evm/blob/5b72bc771ef85750ad7fff056922a137592b38ea/gasometer/src/lib.rs#L175

As to StackExecutor::call_inner and StackExecutor::create_inner, both look correct for me. Let's review them together to double-check.

File: evm_loader/program/src/executor.rs (hunk 5 above)
Comment (vakond) [reply evm_loader/program/src/executor.rs:361]: Added recording of transaction cost in in create_begin & call_begin.

File: evm_loader/program/src/executor.rs (hunk 6 above)
Comment (vakond) [reply evm_loader/program/src/executor.rs]: Added gas recording in all 4 cases: call_begin, create_begin, call & create. Restored the logic from the mainstream rust-blockchain/evm.

File: evm_loader/cli/src/main.rs [hunk 8]
```diff
@@ -416,7 +416,7 @@ fn make_deploy_ethereum_transaction(
         let tx = UnsignedTransaction {
             to: None,
             nonce: trx_count,
-            gas_limit: 1.into(),
+            gas_limit: 9_999_999.into(),
```
Comment (anton-lisanin) [inline evm_loader/cli/src/main.rs:419]: @s-medvedev Shouldn't the gas limit be a configurable in the CLI?

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

File: evm_loader/program/src/entrypoint.rs [hunk 9]
```diff
@@ -456,10 +465,11 @@ fn do_call<'a>(
         if exit_reason.is_succeed() {
             debug_print!("Succeed execution");
             let executor_state = executor.into_state();
+            let used_gas = executor_state.substate().metadata().gasometer().used_gas();
             let (_, (applies, logs)) = executor_state.deconstruct();
-            (exit_reason, result, Some((applies, logs)))
+            (exit_reason, used_gas, result, Some((applies, logs)))
         } else {
-            (exit_reason, result, None)
+            (exit_reason, 0, result, None)
```
[the remark this answers is not in the document]
Comment (vakond) [reply evm_loader/program/src/entrypoint.rs]: Thanks, it's apparent bug. Fixed in two places.

File: evm_loader/program/src/executor.rs [hunk 10]
```diff
@@ -308,28 +368,23 @@ impl<'config, B: Backend> Machine<'config, B> {
...
 
-    pub fn call_begin(&mut self, caller: H160, code_address: H160, input: Vec<u8>, gas_limit: u64) {
-        self.executor.state.inc_nonce(caller);
-
+    pub fn call_begin(&mut self,
+        caller: H160,
+        code_address: H160,
+        input: Vec<u8>,
+        gas_limit: u64,
+    ) -> Result<(), ExitError> {
```
[the remark this answers is not in the document]
Comment (vakond) [reply evm_loader/program/src/executor.rs]: Fixed.

File: evm_loader/program/src/instruction.rs [hunk 11]
```diff
@@ -278,11 +279,16 @@ impl<'a> EvmInstruction<'a> {
 pub fn on_return(
     myself_program_id: &Pubkey,
     status: u8,
+    used_gas: u64,
     result: &[u8]
 ) -> Instruction {
-    let mut data = Vec::with_capacity(2 + result.len());
+    use core::mem;
+
+    let cap = mem::size_of::<u8>() + mem::size_of::<u64>() + result.len();
```
[the remark this answers is not in the document]
Comment (vakond) [reply evm_loader/program/src/instruction.rs]: Thanks for noticing, forgot factor 2. Fixed.

File: evm_loader/program/src/executor_state.rs [hunk 12]
```diff
@@ -17,28 +18,30 @@ struct ExecutorAccount {
     pub reset: bool,
 }
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct ExecutorMetadata {
-    // gasometer: Gasometer<'config>,
+#[derive(Serialize, Deserialize)]
+pub struct ExecutorMetadata<'config> {
+    gasometer: Gasometer<'config>,
```
[the remark this answers is not in the document]
Comment (vakond) [reply evm_loader/program/src/executor_state.rs:23]: Excellent point. Yes, it's seems possible just make const DEFAULT_CONFIG public and use everywhere, eliminating passing copies of config around. I'm checking this possibility right now.

File: evm_loader/cli/src/main.rs [hunk 13]
```diff
@@ -416,7 +416,7 @@ fn make_deploy_ethereum_transaction(
...
             gas_price: 1.into(),
```
[the remark this answers is not in the document]
Comment (s-medvedev) [reply evm_loader/cli/src/main.rs:420]: We will implement work with the gas_price in other issues.

File: evm_loader/cli/src/main.rs (hunk 8 above)
Comment (s-medvedev) [reply evm_loader/cli/src/main.rs:419]: Yes, we will need the ability to set it through an argument when calling the contract deployment. But this should be done as a separate task. Here is another related task - how can the user estimate the required amount of gas?

**Review by s-medvedev — APPROVED** [review]
Please, wait for review from Anton.

File: evm_loader/program/src/entrypoint.rs [hunk 14]
```diff
@@ -481,29 +491,35 @@ fn do_partial_call<'a>(
...
     debug_print!("   caller: {}", account_storage.origin());
     debug_print!(" contract: {}", account_storage.contract());
 
-    executor.call_begin(account_storage.origin(), account_storage.contract(), instruction_data, u64::max_value());
+    executor.call_begin(
+        account_storage.origin(),
+        account_storage.contract(),
+        instruction_data,
+        gas_limit,
+    ).map_err(|_| ProgramError::InvalidInstructionData)?;
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs:515]: `map_err` is redundant
`call_begin` now returns Result<(), ProgramError>

**Review by anton-lisanin — APPROVED** [review]

File: evm_loader/program/src/entrypoint.rs (hunk 14 above)
Comment (vakond) [reply evm_loader/program/src/entrypoint.rs]: Exactly! Fixed in both places.

**Review by anton-lisanin — APPROVED** [review]

**Review by s-medvedev — APPROVED** [review]

**Pull request merged**